tests/phpunit/integration/Core/Modules/REST_Modules_ControllerTest.php

```diff
@@ -0,0 +1,633 @@
+use Google\Site_Kit\Core\Storage\User_Options;
+use Google\Site_Kit\Tests\TestCase;
+use Google\Site_Kit\Tests\FakeHttpClient;
+use WP_REST_Request;
+
+class REST_Modules_ControllerTest extends TestCase {
+
+	public function set_up() {
+		parent::set_up();
+
+		$this->user         = $this->factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		$this->context      = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$this->options      = new Options( $this->context );
+		$this->user_options = new User_Options( $this->context, $this->user->ID );
+		$this->modules      = new Modules( $this->context, $this->options, $this->user_options );
+
+		wp_set_current_user( $this->user->ID );
+
+		// This ensures the REST server is initialized fresh for each test using it.
+		unset( $GLOBALS['wp_rest_server'] );
+		remove_all_filters( 'googlesitekit_rest_routes' );
+	}
+
+	private function setup_fake_module() {
+		$fake_module = new FakeModule( $this->context );
+		$fake_module->set_force_active( true );
+
+		$this->force_set_property( $this->modules, 'modules', array( 'fake-module' => $fake_module ) );
+	}
+
+	public function test_register() {
+		remove_all_filters( 'googlesitekit_apifetch_preload_paths' );
+		$this->modules->register();
+
+		$this->assertTrue( has_filter( 'googlesitekit_apifetch_preload_paths' ) );
+		$this->assertTrue( has_filter( 'googlesitekit_features_request_data' ) );
+		$this->assertContains(
+			'/' . REST_Routes::REST_ROOT . '/core/modules/data/list',
+			apply_filters( 'googlesitekit_apifetch_preload_paths', array() )
+		);
+	}
+
+	public function test_list_rest_endpoint__get_method() {
+		$this->modules->register();
+
+		$request  = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/core/modules/data/list' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertNotEmpty( $response->get_data() );
+	}
+
+	public function test_list_rest_endpoint__no_post_method() {
+		$this->modules->register();
+
+		$request  = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/core/modules/data/list' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 404, $response->get_status() );
+	}
+
+	public function test_list_rest_endpoint__shape() {
+		$this->modules->register();
+
+		$request          = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/core/modules/data/list' );
+		$response         = rest_get_server()->dispatch( $request );
+		$module_data_keys = array(
+			'slug',
+			'name',
+			'description',
+			'homepage',
+			'internal',
+			'order',
+			'forceActive',
+			'shareable',
+			'recoverable',
+			'active',
+			'connected',
+			'dependencies',
+			'dependants',
+			'owner',
+		);
+
+		foreach ( $response->get_data() as $data ) {
+			foreach ( $module_data_keys as $module_data_key ) {
+				$this->assertArrayHasKey( $module_data_key, $data );
```

Comment:
These assertions won't run if the endpoint returns an empty array which could leave this silently unchecked. It's probably unlikely that this could happen with other tests passing but let's add an assertion above to make sure the response is a non-empty array.  We could also check specifically for a modules or two that are guaranteed to be there (either because we set it that way in the test or by checking a `forceActive` module like SC).

Reply:
Test updated 👍